p2panda-spaces/src/space.rs

```diff
@@ -374,6 +490,89 @@ where
         Ok(events)
     }
 
+    /// Process an auth control message before the "authored" version has been forged. This is
+    /// useful when we want to know what the resulting auth state will be.
+    async fn pre_process_auth(
+        manager: Manager<S, F, M, C, RS>,
+        space_id: ActorId,
+        dependencies: &Vec<OperationId>,
+        control_message: &AuthControlMessage<C>,
+    ) -> Result<(Vec<ActorId>, Vec<ActorId>), SpaceError<S, F, M, C, RS>> {
+        let manager = manager.inner.read().await;
+        let my_id: ActorId = manager.forge.public_key().into();
+        let auth_y = manager.store.auth().await.map_err(SpaceError::AuthStore)?;
+        let previous_members = secret_members(auth_y.members(space_id));
+
+        // We process a fake operation on the current auth state in order to compute the next
+        // membership state.
+        let fake_op = AuthMessage::Forged {
+            author: my_id,
+            operation_id: OperationId::placeholder(),
+            args: AuthArgs {
+                dependencies: dependencies.to_owned(),
+                control_message: control_message.to_owned(),
+            },
+        };
+
+        // NOTE: as we are only calling this method when we ourselves want to make space
+        // membership changes locally, no operations will be concurrent to our local auth
+        // state, and so no re-build will occur. This means processing the operation is cheap.
+        // Processing the operation through the auth api is preferred over calculating the
+        // next state manually, as auth does as some basic validation of control messages for
+        // us.
+        let auth_y_i = AuthGroup::process(auth_y, &fake_op).map_err(SpaceError::AuthGroup)?;
+        let next_members = secret_members(auth_y_i.members(space_id));
+        let diff_members = next_members
+            .iter()
+            .cloned()
+            .filter(|actor| previous_members.contains(actor))
+            .collect::<Vec<_>>();
+        Ok((next_members, diff_members))
+    }
```

Comment:
Cool that this was so easy to do in the end. For performance we could also later do an even simpler approach where we take the current transitive members and add / remove / promote / demote them manually. Since this is a local operation we don't need to check for concurrency.

Reply:
I don't think there's a performance consideration here, as if an operation isn't concurrent (it's state points at the current tip) then no heavy processing kicks in (the rebuild logic). I think using the existing API here is nice, it keeps the validation consistent and in one (already well considered) place and adding a nested group is handled already (we'd have to query each added/removed group and resolve transitive members manually).

Reply:
Ah great! This sounds good indeed!